doozerlib/repodata.py

```diff
@@ -0,0 +1,314 @@
+        if not url:
+            return b''
+        data = io.BytesIO()
+        async with session.get(url) as resp:
+            resp.raise_for_status()
+            data = io.BytesIO(await resp.read())
+        data.seek(0)
+        with gzip.GzipFile(fileobj=data) as uncompressed:
+            return uncompressed.read()
+
+    async def load(self, repo_name: str, repo_url: str):
+        if not repo_url.endswith("/"):
+            repo_url += "/"
+        repomd_url = parse.urljoin(repo_url, "repodata/repomd.xml")
+
+        timeout = aiohttp.ClientTimeout(total=60 * 10)
+        async with aiohttp.ClientSession(connector=aiohttp.TCPConnector(limit=32, force_close=True), timeout=timeout) as session:
+            async with session.get(repomd_url) as resp:
+                resp.raise_for_status()
+                repomd_xml = ET.fromstring(await resp.text())
+
+            primary_data_element = repomd_xml.find('repo:data[@type="primary"]', NAMESPACES)
+            if primary_data_element is None:
+                raise ValueError("Couldn't find primary data in repodata")
+            primary_location = primary_data_element.find('repo:location', NAMESPACES)
+            if primary_location is None:
+                raise ValueError("Couldn't find primary location in repodata")
+            primary_url = parse.urljoin(repo_url, primary_location.attrib['href'])
+
+            modules_url = None
+            modules_data_element = repomd_xml.find('repo:data[@type="modules"]', NAMESPACES)
+            if modules_data_element is not None:
+                modules_location = modules_data_element.find('repo:location', NAMESPACES)
+                if modules_location is None:
+                    raise ValueError("Couldn't find modules location in repodata")
+                modules_url = parse.urljoin(repo_url, modules_location.attrib['href'])
+
+            @retry(reraise=True, stop=stop_after_attempt(5), wait=wait_exponential(multiplier=1, min=1, max=10),
+                   retry=(retry_if_exception_type((aiohttp.ServerDisconnectedError, aiohttp.ClientResponseError, aiohttp.ClientPayloadError))),
+                   before_sleep=before_sleep_log(LOGGER, logging.WARNING))
+            async def fetch_remote_gzip(url: Optional[str]):
+                return await self._fetch_remote_gzip(session, url)
+
+            primary_bytes, modules_bytes = await asyncio.gather(
+                fetch_remote_gzip(primary_url),
+                fetch_remote_gzip(modules_url),
+            )
+
+        yaml = YAML(typ='safe')
+        repodata = Repodata.from_metadatas(
+            repo_name,
+            ET.fromstring(primary_bytes),
+            list(yaml.load_all(modules_bytes) if modules_bytes else [])
+        )
+        return repodata
+
+
+class OutdatedRPMFinder:
+    async def find_non_latest_rpms(self, rpms_to_check: List[Dict], repodatas: List[Repodata], logger: Optional[Logger] = None) -> List[Tuple[str, str, str]]:
```

Comment:
looong method. looks like it would be easy to break up into smaller testable methods?... i mean you managed to test it anyway, just seems like it would've been cleaner broken out.